option optimize_for = LITE_RUNTIME;

message Data {

    oneof data_type {
        FeatureCollection feature_collection = 1;
        Feature feature = 2;
        Geometry geometry = 3;
    }

    // global arrays of unique keys
    repeated string keys = 4;

    optional uint32 dimensions = 6 [default = 2]; // max coordinate dimensions

    // number of digits after decimal point for coordinates
    optional uint32 precision = 7 [default = 6];

    // TopoJSON stuff
    optional Transform transform = 8;
    repeated sint32 arc_coords = 9 [packed = true];
    repeated uint32 arc_lengths = 10 [packed = true];

    repeated Value values = 14; // unique values
    repeated uint32 custom_properties = 15 [packed = true]; // arbitrary properties


    message Feature {
        required Geometry geometry = 1;

        repeated uint32 properties = 2 [packed = true]; // pairs of key/value indexes
        oneof id_type {
            string id = 3;
            sint32 int_id = 4;
        }

        repeated Value values = 14;
        repeated uint32 custom_properties = 15 [packed = true];
    }

    message Geometry {
        required Type type = 1;

        // TopoJSON has props/id in Geometry rather than Feature
        repeated uint32 properties = 2 [packed = true];
        oneof id_type {
            string id = 3;
            sint32 int_id = 4;
        }

        repeated sint32 coords = 6 [packed = true]; // delta-encoded integer values
        repeated uint32 lengths = 7 [packed = true]; // coordinate structure in lengths
        repeated Geometry geometries = 8;

        optional string name = 9; // TopoJSON object name

        repeated Value values = 14;
        repeated uint32 custom_properties = 15 [packed = true];
    }

    message FeatureCollection {
        repeated Feature features = 1;

        repeated Value values = 14;
        repeated uint32 custom_properties = 15 [packed = true];
    }

    enum Type {
        POINT = 0;
        MULTIPOINT = 1;
        LINESTRING = 2;
        MULTILINESTRING = 3;
        POLYGON = 4;
        MULTIPOLYGON = 5;
        GEOMETRYCOLLECTION = 6;
    }

    message Value {
        oneof value_type {
            string string_value = 1;
            double double_value = 2;
            sint32 int_value = 3;
            uint32 uint_value = 4;
            bool bool_value = 5;
            string json_value = 6;
        }
    }

    message Transform {
        required double scale_x = 1;
        required double scale_y = 2;
        required double translate_x = 3;
        required double translate_y = 4;
    }
}
